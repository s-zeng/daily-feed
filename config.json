{
  "sources": [
    {
      "type": "ars_technica",
      "name": "Ars Technica"
    },
    {
      "type": "hackernews",
      "name": "Hacker News"
    },
    {
      "type": "rss",
      "name": "BBC News - Technology",
      "url": "http://feeds.bbci.co.uk/news/technology/rss.xml",
      "description": "BBC technology news"
    },
    {
      "type": "rss",
      "name": "Bloomberg - Technology",
      "url": "https://feeds.bloomberg.com/technology/news.rss",
      "description": "Bloomberg technology news"
    }
  ],
  "output": {
    "filename": "daily-feed.epub",
    "title": "Daily Feed Digest",
    "author": "RSS Aggregator"
  },
  "front_page": {
    "enabled": true,
    "provider": {
      "type": "ollama",
      "base_url": "http://127.0.0.1:1234",
      "model": "google/gemma-3-27b"
    }
  }
}
