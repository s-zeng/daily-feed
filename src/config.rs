use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fs;
use crate::ai_client::AiProvider;

#[derive(Debug, Deserialize, Serialize)]
pub struct Feed {
    pub name: String,
    pub url: String,
    pub description: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum OutputFormat {
    #[serde(rename = "epub")]
    Epub,
    #[serde(rename = "markdown")]
    Markdown,
}

impl Default for OutputFormat {
    fn default() -> Self {
        OutputFormat::Epub
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct FrontPageConfig {
    pub enabled: bool,
    pub provider: AiProviderConfig,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "type")]
pub enum AiProviderConfig {
    #[serde(rename = "ollama")]
    Ollama {
        base_url: String,
        model: String,
    },
    #[serde(rename = "anthropic")]
    Anthropic {
        api_key: String,
        model: String,
    },
}

impl From<AiProviderConfig> for AiProvider {
    fn from(config: AiProviderConfig) -> Self {
        match config {
            AiProviderConfig::Ollama { base_url, model } => AiProvider::Ollama { base_url, model },
            AiProviderConfig::Anthropic { api_key, model } => AiProvider::Anthropic { api_key, model },
        }
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct OutputConfig {
    pub filename: String,
    pub title: String,
    pub author: String,
    #[serde(default)]
    pub format: OutputFormat,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Config {
    pub feeds: Vec<Feed>,
    pub output: OutputConfig,
    pub front_page: Option<FrontPageConfig>,
}

impl Config {
    pub fn load_from_file(path: &str) -> Result<Self, Box<dyn Error>> {
        let content = fs::read_to_string(path)?;
        let config: Config = serde_json::from_str(&content)?;
        Ok(config)
    }

    pub fn default() -> Self {
        Config {
            feeds: vec![Feed {
                name: "Ars Technica".to_string(),
                url: "https://arstechnica.com/feed/?t=bd10cf5f81b5d0a5edbb2faa32e6d55c7a1efdb3"
                    .to_string(),
                description: "Technology news and insights".to_string(),
            }],
            output: OutputConfig {
                filename: "daily-feed.epub".to_string(),
                title: "Daily Feed Digest".to_string(),
                author: "RSS Aggregator".to_string(),
                format: OutputFormat::default(),
            },
            front_page: None,
        }
    }
}
