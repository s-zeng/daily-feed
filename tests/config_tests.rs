use daily_feed::config::{Config, Feed, OutputConfig, OutputFormat};
use std::fs;
use tempfile::NamedTempFile;

#[test]
fn test_config_deserialization() {
    let json_content = r#"
    {
        "feeds": [
            {
                "type": "generic",
                "name": "Test Feed",
                "url": "https://test.example.com/feed.xml",
                "description": "A test feed"
            }
        ],
        "output": {
            "filename": "test_output.epub",
            "title": "Test Title",
            "author": "Test Author"
        }
    }
    "#;

    let config: Config = serde_json::from_str(json_content).unwrap();
    insta::assert_json_snapshot!(config);
}

#[test]
fn test_config_load_from_file() {
    let json_content = r#"
    {
        "feeds": [
            {
                "type": "generic",
                "name": "File Test Feed",
                "url": "https://file.example.com/feed.xml",
                "description": "A test feed from file"
            }
        ],
        "output": {
            "filename": "file_test.epub",
            "title": "File Test Title",
            "author": "File Test Author"
        }
    }
    "#;

    let temp_file = NamedTempFile::new().unwrap();
    fs::write(temp_file.path(), json_content).unwrap();

    let config = Config::load_from_file(temp_file.path().to_str().unwrap()).unwrap();
    insta::assert_json_snapshot!(config);
}

#[test]
fn test_config_load_from_nonexistent_file() {
    let result = Config::load_from_file("nonexistent.json");
    let error_message = result.unwrap_err().to_string();
    insta::assert_snapshot!(error_message);
}

#[test]
fn test_config_load_invalid_json() {
    let invalid_json = r#"
    {
        "feeds": [
            {
                "name": "Test Feed",
                "url": "https://test.example.com/feed.xml"
                // Missing description and closing brace
    "#;

    let temp_file = NamedTempFile::new().unwrap();
    fs::write(temp_file.path(), invalid_json).unwrap();

    let result = Config::load_from_file(temp_file.path().to_str().unwrap());
    let error_message = result.unwrap_err().to_string();
    insta::assert_snapshot!(error_message);
}

#[test]
fn test_config_default() {
    let config = Config::default();
    insta::assert_json_snapshot!(config);
}

#[test]
fn test_config_serialization() {
    let config = Config {
        feeds: vec![Feed::Generic {
            name: "Serialize Test".to_string(),
            url: "https://serialize.example.com/feed.xml".to_string(),
            description: "Test serialization".to_string(),
        }],
        output: OutputConfig {
            filename: "serialize_test.epub".to_string(),
            title: "Serialize Test Title".to_string(),
            author: "Serialize Test Author".to_string(),
            format: OutputFormat::Epub,
        },
        front_page: None,
    };

    let json = serde_json::to_string(&config).unwrap();
    let deserialized: Config = serde_json::from_str(&json).unwrap();
    insta::assert_json_snapshot!(deserialized);
}

#[test]
fn test_config_multiple_feeds() {
    let json_content = r#"
    {
        "feeds": [
            {
                "type": "generic",
                "name": "Feed 1",
                "url": "https://feed1.example.com/feed.xml",
                "description": "First feed"
            },
            {
                "type": "generic",
                "name": "Feed 2",
                "url": "https://feed2.example.com/feed.xml",
                "description": "Second feed"
            },
            {
                "type": "ars_technica"
            }
        ],
        "output": {
            "filename": "multi_feed_test.epub",
            "title": "Multi Feed Test",
            "author": "Multi Feed Author"
        }
    }
    "#;

    let config: Config = serde_json::from_str(json_content).unwrap();
    insta::assert_json_snapshot!(config);
}
