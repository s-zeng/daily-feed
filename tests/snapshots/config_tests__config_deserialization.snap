---
source: tests/config_tests.rs
expression: config
---
{
  "sources": [],
  "feeds": [
    {
      "type": "generic",
      "name": "Test Feed",
      "url": "https://test.example.com/feed.xml",
      "description": "A test feed"
    }
  ],
  "output": {
    "filename": "test_output.epub",
    "title": "Test Title",
    "author": "Test Author",
    "format": "epub"
  },
  "front_page": null
}
